src/map/trade.cpp

```diff
@@ -413,10 +413,23 @@ void trade_tradeadditem(map_session_data *sd, int16 index, int16 amount)
 	// Locate a trade position
 	ARR_FIND( 0, 10, trade_i, sd->deal.item[trade_i].index == index || sd->deal.item[trade_i].amount == 0 );
 	if( trade_i == 10 ) { // No space left
-		// clif_tradeitemok(*sd, index, EXITEM_ADD_FAILED_OVERWEIGHT); // We do not know if the server respond with this or not since the official client prevents this case client-side.
 		return;
 	}
 
+	// Fail to add the item if the inventory is full
+	if (pc_inventoryblank(target_sd) <= trade_i) {
+		clif_tradeitemok(*sd, index, EXITEM_ADD_FAILED_OVERCOUNT);
+		return;
+	}
+
+	// Fail to add the item if is stacked over the limit
+	if (itemdb_isstackable(item->nameid)) {
```

Comment:
I think to really check this correctly we would need to add a new variable to sd->deal with the required slots.
Here you only calculate the required slots for the current item, but it may be necessary to calculate the required slots for ALL the items in the deal.
See clif_parse_Mail_getattach for a reference.

Reply:
That's not how it works, both on officials and rathena it's checked each time you add a new item to the trade, this checks if the other player will exceed 30k of the item with the amount you tried to trade, otherwise is not added to the list and sends the clif_tradeitemok.
And you can't really re-add or update an item to the trade list, we just have the code below (the sd->deal.item checks) to avoid possible exploits.